### PR TITLE
updated toolkit to use Serverless Cosmos DB by default, configrable f…

### PR DESCRIPTION
**Pull request by jcodella**
Serverless Cosmos DB by defeatul, , configrable for autoscale (1000 RU/s default, user-configurable)

### CODE REVIEW

File: Docs/local_testing.md [hunk 1]
```diff
@@ -153,6 +161,10 @@ memory = AgentMemory(
     cosmos_endpoint=os.getenv("COSMOS_DB_ENDPOINT"),
     cosmos_database=os.getenv("COSMOS_DB_DATABASE"),
     cosmos_container=os.getenv("COSMOS_DB_CONTAINER"),
+    cosmos_counter_container=os.getenv("COSMOS_DB_COUNTERS_CONTAINER", "counter"),
+    cosmos_lease_container=os.getenv("COSMOS_DB_LEASE_CONTAINER", "leases"),
+    cosmos_throughput_mode=os.getenv("COSMOS_DB_THROUGHPUT_MODE", "serverless"),
+    cosmos_autoscale_max_ru=int(os.getenv("COSMOS_DB_AUTOSCALE_MAX_RU", "1000")),
     ai_foundry_endpoint=os.getenv("AI_FOUNDRY_ENDPOINT"),
```
[the remark this answers is not in the document]
Comment (jcodella) [reply Docs/local_testing.md:168]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3118618307)

File: Docs/azure_testing.md [hunk 2]
```diff
@@ -192,6 +203,10 @@ memory = AgentMemory(
     cosmos_endpoint=os.getenv("COSMOS_DB_ENDPOINT"),
     cosmos_database=os.getenv("COSMOS_DB_DATABASE"),
     cosmos_container=os.getenv("COSMOS_DB_CONTAINER"),
+    cosmos_counter_container=os.getenv("COSMOS_DB_COUNTERS_CONTAINER", "counter"),
+    cosmos_lease_container=os.getenv("COSMOS_DB_LEASE_CONTAINER", "leases"),
+    cosmos_throughput_mode=os.getenv("COSMOS_DB_THROUGHPUT_MODE", "serverless"),
+    cosmos_autoscale_max_ru=int(os.getenv("COSMOS_DB_AUTOSCALE_MAX_RU", "1000")),
     ai_foundry_endpoint=os.getenv("AI_FOUNDRY_ENDPOINT"),
```
[the remark this answers is not in the document]
Comment (jcodella) [reply Docs/azure_testing.md:210]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3118618331)

File: Docs/local_testing.md [hunk 3]
```diff
@@ -85,13 +88,18 @@ ADF_KEY=
 
 The Functions runtime uses `azure_functions/local.settings.json`, not `.env`, so mirror the same values there.
 
+`COSMOS_DB_THROUGHPUT_MODE=serverless` is the default and creates the required Cosmos containers without specifying RU/s. If you set `COSMOS_DB_THROUGHPUT_MODE=autoscale`, the toolkit provisions the memories, counter, and lease containers with the shared max RU/s value from `COSMOS_DB_AUTOSCALE_MAX_RU`.
+
```
[the remark this answers is not in the document]
Comment (jcodella) [reply Docs/local_testing.md:92]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3118618362)

File: Samples/Demo_async.ipynb [hunk 4]
```diff
@@ -74,20 +113,46 @@
...
+    "\n",
+    "    cosmos_counter_container=os.getenv(\"COSMOS_DB_COUNTERS_CONTAINER\", \"counter\"),\n",
+    "\n",
+    "    cosmos_lease_container=os.getenv(\"COSMOS_DB_LEASE_CONTAINER\", \"leases\"),\n",
+    "\n",
+    "    cosmos_throughput_mode=os.getenv(\"COSMOS_DB_THROUGHPUT_MODE\", \"serverless\"),\n",
+    "\n",
+    "    cosmos_autoscale_max_ru=int(os.getenv(\"COSMOS_DB_AUTOSCALE_MAX_RU\", \"1000\")),\n",
+    "\n",
     "    ai_foundry_endpoint=os.getenv(\"AI_FOUNDRY_ENDPOINT\"),\n",
```
[the remark this answers is not in the document]
Comment (jcodella) [reply Samples/Demo_async.ipynb:136]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3118618287)

File: Samples/Demo.ipynb [hunk 5]
```diff
@@ -76,19 +118,43 @@
...
     "    cosmos_container=os.getenv(\"COSMOS_DB_CONTAINER\"),\n",
+    "\n",
+    "    cosmos_counter_container=os.getenv(\"COSMOS_DB_COUNTERS_CONTAINER\", \"counter\"),\n",
+    "\n",
+    "    cosmos_lease_container=os.getenv(\"COSMOS_DB_LEASE_CONTAINER\", \"leases\"),\n",
+    "\n",
+    "    cosmos_throughput_mode=os.getenv(\"COSMOS_DB_THROUGHPUT_MODE\", \"serverless\"),\n",
+    "\n",
+    "    cosmos_autoscale_max_ru=int(os.getenv(\"COSMOS_DB_AUTOSCALE_MAX_RU\", \"1000\")),\n",
+    "\n",
```
[the remark this answers is not in the document]
Comment (jcodella) [reply Samples/Demo.ipynb:137]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3118618130)

File: README.md [hunk 6]
```diff
@@ -134,14 +134,20 @@ memory = CosmosMemoryClient(
     cosmos_endpoint=os.getenv("COSMOS_DB_ENDPOINT"),
     cosmos_database=os.getenv("COSMOS_DB_DATABASE"),
     cosmos_container=os.getenv("COSMOS_DB_CONTAINER"),
+    cosmos_counter_container=os.getenv("COSMOS_DB_COUNTERS_CONTAINER", "counter"),
+    cosmos_lease_container=os.getenv("COSMOS_DB_LEASE_CONTAINER", "leases"),
+    cosmos_throughput_mode=os.getenv("COSMOS_DB_THROUGHPUT_MODE", "serverless"),
+    cosmos_autoscale_max_ru=int(os.getenv("COSMOS_DB_AUTOSCALE_MAX_RU", "1000")),
     ai_foundry_endpoint=os.getenv("AI_FOUNDRY_ENDPOINT"),
```
[the remark this answers is not in the document]
Comment (jcodella) [reply README.md:141]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3118618086)

File: tests/unit/aio/test_cosmos_memory_client.py [hunk 7]
```diff
@@ -262,21 +264,62 @@ async def test_create_memory_store_with_counter_container(self):
...
+                    PartitionKey=MagicMock(),
+                    ThroughputProperties=MagicMock(),
+                ),
+            },
+        ):
+            await mem.create_memory_store(
+                endpoint="https://fake.documents.azure.com:443/",
+                credential="fake-key",
+            )
+
```
[the remark this answers is not in the document]
Comment (jcodella) [reply tests/unit/aio/test_cosmos_memory_client.py:317]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3112211541)

File: tests/unit/test_cosmos_memory_client.py [hunk 8]
```diff
@@ -279,21 +284,64 @@ def test_create_memory_store_with_custom_dimensions(self):
...
+                    PartitionKey=MagicMock(),
+                    ThroughputProperties=MagicMock(),
+                ),
+            },
+        ):
+            mem.create_memory_store(
+                endpoint="https://fake.documents.azure.com:443/",
+                credential="fake-key",
+            )
+
```
[the remark this answers is not in the document]
Comment (jcodella) [reply tests/unit/test_cosmos_memory_client.py:339]: @copilot apply changes based on [this feedback](https://github.com/AzureCosmosDB/AgentMemoryToolkit/pull/6#discussion_r3112211520)

File: agent_memory_toolkit/_utils.py [hunk 9]
```diff
@@ -69,6 +69,70 @@ def _resolve_embedding_dimensions(val: Optional[int]) -> Optional[int]:
...
+            ),
+            parameter="cosmos_throughput_mode",
+        )
+    return raw
+
+
+def _resolve_cosmos_autoscale_max_ru(val: Optional[int]) -> int:
+    """Resolve autoscale max RU from explicit value or env var."""
+    if val is not None:
+        return val
```
Comment (aayush3011) [inline agent_memory_toolkit/_utils.py:97]: 🔴 **Bug: Validation is bypassed for explicitly-passed values**

When a caller passes `cosmos_autoscale_max_ru=0` or `cosmos_autoscale_max_ru=-1` explicitly, this early return fires before the `<= 0` validation check (which only runs for env-var-parsed values at line 100). The invalid value flows through to `ThroughputProperties(auto_scale_max_throughput=0)`, causing a cryptic Cosmos SDK error instead of the intended `ConfigurationError`.

```python
# Reproduces the bug:
client = CosmosMemoryClient(
    cosmos_throughput_mode="autoscale",
    cosmos_autoscale_max_ru=0,  # Invalid but accepted
)
# → _resolve_cosmos_autoscale_max_ru(0) returns 0
# → ThroughputProperties(auto_scale_max_throughput=0) → Azure SDK error
```

**Suggested fix:**
```python
if val is not None:
    if not isinstance(val, int) or isinstance(val, bool) or val <= 0:
        raise ConfigurationError(
            message=f"Invalid configuration for cosmos_autoscale_max_ru: expected a positive integer, got '{val}'",
            parameter="cosmos_autoscale_max_ru",
        )
    return val
```

File: agent_memory_toolkit/_utils.py [hunk 10]
```diff
@@ -69,6 +69,70 @@ def _resolve_embedding_dimensions(val: Optional[int]) -> Optional[int]:
     return parsed if parsed else None
 
 
+def _resolve_cosmos_throughput_mode(val: Optional[str]) -> str:
+    """Resolve throughput mode from explicit value or env var.
+
+    Allowed values are ``serverless`` and ``autoscale``.
+    """
+    raw = (val or os.environ.get("COSMOS_DB_THROUGHPUT_MODE") or "serverless").strip().lower()
```
Comment (aayush3011) [inline agent_memory_toolkit/_utils.py]: 🟡 **Truthy `or` chain silently swallows empty-string misconfigurations**

Python's `or` treats `""` as falsy. If a user passes `cosmos_throughput_mode=""`, it silently falls through to the env var or default `"serverless"` instead of raising a `ConfigurationError`. An empty string is almost certainly a misconfiguration — the user expects it to mean something, not to be quietly replaced with a default.

The same pattern exists in `create_memory_store()` where `throughput_mode or self._cosmos_throughput_mode` silently drops `""`.

**Suggested fix** — use explicit `None` checks:
```python
raw = (val if val is not None else os.environ.get("COSMOS_DB_THROUGHPUT_MODE") or "serverless").strip().lower()
```

And in both `create_memory_store()` implementations:
```python
self._cosmos_throughput_mode = _resolve_cosmos_throughput_mode(
    throughput_mode if throughput_mode is not None else self._cosmos_throughput_mode
)
```

File: agent_memory_toolkit/cosmos_memory_client.py [hunk 11]
```diff
@@ -382,40 +405,56 @@ def create_memory_store(
...
+                counter_kwargs["offer_throughput"] = offer_throughput
+            db.create_container_if_not_exists(**counter_kwargs)
+
+            lease_kwargs = {
+                "id": self._cosmos_lease_container,
+                "partition_key": lease_partition_key,
+            }
+            if offer_throughput is not None:
+                lease_kwargs["offer_throughput"] = offer_throughput
+            db.create_container_if_not_exists(**lease_kwargs)
```
Comment (aayush3011) [inline agent_memory_toolkit/cosmos_memory_client.py]: 🟢 **Suggestion: Extract a helper to reduce container-kwargs repetition**

This `build dict → conditionally add offer_throughput → create_container_if_not_exists` pattern is repeated 3 times here and 3 times in the async client (6 total). If a new container-level property needs to be added later (e.g., TTL policy, analytical store), it must be updated in 6 places.

Since `_utils.py` already centralizes other provisioning logic, consider extracting a helper:

```python
# In _utils.py:
def _build_container_kwargs(container_id, partition_key, offer_throughput, **extras):
    kwargs = {"id": container_id, "partition_key": partition_key, **extras}
    if offer_throughput is not None:
        kwargs["offer_throughput"] = offer_throughput
    return kwargs
```

Each call site then becomes a one-liner:
```python
container_handle = db.create_container_if_not_exists(
    **_build_container_kwargs(self._cosmos_container, partition_key, offer_throughput,
                              indexing_policy=idx_policy, vector_embedding_policy=vec_policy,
                              full_text_policy=ft_policy)
)
```

**Review by aayush3011 — CHANGES_REQUESTED** [review]
## Review Summary

The architecture is solid — centralizing throughput configuration into `_utils.py` and taking ownership of all three containers is the right move. A few issues to address before merging:

### General findings (not tied to a specific diff line)

**🟡 `function_app.py` docstring contradicts updated documentation**

Every documentation file in this PR now says the leases container is *"provisioned by `create_memory_store()`"* (concepts.md, azure_testing.md, local_testing.md, README.md). However, `azure_functions/function_app.py` line 54 still says:

```python
- ``leases``   – auto-created by the trigger for change feed checkpointing
```

This should be updated to reflect the new provisioning model, e.g.:
```python
- ``leases``   – provisioned by create_memory_store(); trigger fallback-creates if missing
```

---

**🟡 No dedicated unit tests for the four new `_utils.py` functions**

The four new functions (`_resolve_cosmos_throughput_mode`, `_resolve_cosmos_autoscale_max_ru`, `_resolve_cosmos_provisioning_autoscale_max_ru`, `_cosmos_container_offer_throughput`) are only tested indirectly through client-level tests. A dedicated `tests/unit/test_utils.py` would make edge-case coverage much more robust. For example, the validation bypass in the inline comment below (Finding 1) would have been caught by a simple:

```python
def test_resolve_autoscale_max_ru_rejects_zero():
    with pytest.raises(ConfigurationError):
        _resolve_cosmos_autoscale_max_ru(0)

def test_resolve_autoscale_max_ru_rejects_negative():
    with pytest.raises(ConfigurationError):
        _resolve_cosmos_autoscale_max_ru(-1)

def test_resolve_autoscale_max_ru_rejects_bool():
    with pytest.raises(ConfigurationError):
        _resolve_cosmos_autoscale_max_ru(True)
```

File: agent_memory_toolkit/cosmos_memory_client.py (hunk 11 above)
Comment (jcodella) [reply agent_memory_toolkit/cosmos_memory_client.py]: Adding in next commit

File: agent_memory_toolkit/_utils.py (hunk 10 above)
Comment (jcodella) [reply agent_memory_toolkit/_utils.py]: done in next commit

File: agent_memory_toolkit/_utils.py (hunk 9 above)
Comment (jcodella) [reply agent_memory_toolkit/_utils.py:97]: Done in next commit.

**Review by aayush3011 — APPROVED** [review]
